mabwiser/popularity.py

```diff
@@ -0,0 +1,48 @@
+# -*- coding: utf-8 -*-
+# SPDX-License-Identifer: Apache-2.0
+
+from typing import List, Optional, NoReturn
+import numpy as np
+
+from mabwiser.greedy import _EpsilonGreedy
+from mabwiser.utils import Arm, reset
+
+
+class _Popularity(_EpsilonGreedy):
+
+    def __init__(self, rng: np.random.RandomState, arms: List[Arm], n_jobs: int, backend: Optional[str]):
+
+        # Init the parent greedy policy with zero epsilon
+        super().__init__(rng, arms, n_jobs, backend, epsilon=0.0)
+
+    def fit(self, decisions: np.ndarray, rewards: np.ndarray, contexts: np.ndarray = None) -> NoReturn:
+
+        # Fit as usual greedy
+        super().fit(decisions, rewards, contexts)
+
+        # Make sure expectations sum up to 1 like probabilities
+        self._normalize_expectations()
+
+    def partial_fit(self, decisions: np.ndarray, rewards: np.ndarray, contexts: np.ndarray = None) -> NoReturn:
+
+        # Fit as usual greedy
+        super().partial_fit(decisions, rewards, contexts)
+
+        # Make sure expectations sum up to 1 like probabilities
+        self._normalize_expectations()
+
+    def predict(self, contexts: np.ndarray = None) -> Arm:
+
+        # Select an arm randomized by expectation probability
+        return self.rng.choice(self.arms, p=list(self.arm_to_expectation.values()))
+
+    def _normalize_expectations(self):
+        # TODO: this would not work for negative rewards!
```

Comment:
no, I don't plan to address this as we don't have an immediate case for it, but it's a warning in case someone comes here. We shall address it though eventually.

Reply:
Btw, I don't know what that algorithm for the case of negative rewards should look like. It's not obvious to me.. Maybe shift everything by min(rewards) first, and scale later.. Or see the L1 or L2 versions.. Each will probably have different behaviour/performanece and it's not straightforward to me how that should it look like. Something to think in the background

Reply:
Shifting by the min makes sense. Perhaps this TODO should be moved to be a line in the API documentation? Also, it might be good to add a unit test to verify what the behavior is when some of the arms have negative means